### ISSUE TITLE
Add option `HUNTER_INSTALL_WITH_DEPENDENCIES` to install dependencies to `CMAKE_INSTALL_PREFIX`

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#274]
### Alternative approaches
* Copy all files from `HUNTER_INSTALL_PREFIX` location to the destination directory (add target/install code?). Not an optimal way because may copy dependencies installed by other projects since this directory is shared.
* Same as above but choose some new `HUNTER_ROOT` directory only for this project, i.e. build everything from scratch. Not optimal because `HUNTER_ROOT` is not shared.
* Just use Hunter! :)

### Approaches that will not work
* We can't just change final destination (`CMAKE_PREFIX_PATH` or `./configure --prefix`) of Hunter packages since some service information saved in `HUNTER_ROOT`. E.g. we install `Boost` to `/custom/user/install/location` and save `DONE` stamp. Next project that want to see `Boost` will not install anything because `DONE` is saved. At the same time it doesn't know anything about `/custom/user/install/location`.
* Since not all projects use CMake as a build system we can't run build twice with different `CMAKE_PREFIX_PATH`: `cmake -DCMAKE_PREFIX_PATH=$HUNTER_INSTALL_PREFIX` and `cmake -DCMAKE_PREFIX_PATH=/custom/user/install/location`.
* Even if we can suggest that two different install destination can be used, packages can be installed from cache hence need to update "unpack from cache" logic (+ dependencies).

### Implementation

* Files from Hunter root (libraries/headers) will be installed by command `install(SCRIPT ...)`. This command will be added on Hunter initialization in the end of [hunter_finalize](https://github.com/ruslo/hunter/blob/master/cmake/modules/hunter_finalize.cmake) module (after we know `HUNTER_INSTALL_PREFIX`). And only if `HUNTER_INSTALL_WITH_DEPENDENCIES` is `ON`. Snippet will looks like this:
```cmake
if(HUNTER_INSTALL_WITH_DEPENDENCIES)
  hunter_test_string_not_empty("${CMAKE_BINARY_DIR}") # sanity check
  set(
      HUNTER_INSTALL_DEPENDENCIES_WORKING_DIRECTORY
      "${CMAKE_BINARY_DIR}/_3rdParty/Hunter/install-dependencies"
  )
  file(MAKE_DIRECTORY "${HUNTER_INSTALL_DEPENDENCIES_WORKING_DIRECTORY}")
  configure_file(
      "${HUNTER_SELF}/cmake/install-dependencies.cmake.in"
      "${HUNTER_INSTALL_DEPENDENCIES_WORKING_DIRECTORY}/run.cmake"
      @ONLY
  )
  install(SCRIPT "${HUNTER_INSTALL_DEPENDENCIES_WORKING_DIRECTORY}/run.cmake")
endif()
```
* Each project should save list of installed files
* Each `hunter_download` command should copy information about package and it's dependencies (see [hunter_get_dependency_list](https://github.com/ruslo/hunter/blob/master/cmake/modules/hunter_get_dependency_list.cmake)) to the `HUNTER_INSTALL_DEPENDENCIES_WORKING_DIRECTORY` (must be visible globally in this case, cache variable?). Format:
```
- packages/
    <package>/
      - <package>.list
      __components/
        <component>.list
```
* script `run.cmake` (template `run.cmake.in`) must use `*.list` files to install it to the `CMAKE_INSTALL_PREFIX` directory

### Notes

* does `install(SCRIPT` code need to be protected with guard to avoid double execution? If so we can't use cache variables (will be `TRUE` if rerun from scratch) and can't use regular variables (not visible from different directories, see PARENT_SCOPE) => need to add global property variable.
* We need to copy custom `Find*.cmake` files that used from Hunter (https://github.com/ruslo/hunter/tree/master/cmake/find). This files used as a workaround/improvements of standard CMake modules. We can install it too but user have to modify `CMAKE_MODULE_PATH` variable manually to see them.

**Comment by ruslo** [comment]
By using the `GIT_SELF` feature effectively functionality is the same as requested: we creating separate directory with installed dependencies and installed package itself:
* https://docs.hunter.sh/en/latest/user-guides/hunter-user/git-self.html

**Comment by ruslo** [comment]
Workarounds:

* If local library is shared you can build dependent libraries as static. In this case all symbols from static libraries will be absorbed by shared library and you don't have to install them at all.

* If all libraries are shared or all libraries are static you can install dependent libraries manually using pattern:
```cmake
install(
    FILES
    $<TARGET_FILE:OpenSSL::Crypto>
    $<TARGET_FILE:ZLIB::zlib>
    DESTINATION
    lib
)
```

**Comment by ithron** [comment]
I see a general problem here:
let's say I'm building a library MyLib. And my public headers are depending on Eigen3 headers.
Then Eigen3 public headers need to be installed when MyLib is installed. But somehow one must ensure that the Eigen3 headers are installed into a private prefix (e.g. /usr/local/include/MyLib/private/Eigen) so that it does not conflict with other installations of Eigen.
However, it MyLib does only privately (e.g. target_link_libraries(MyLib PRIVATE Eigen3::Eigen) ) then the Eigen3 don't need to be installed at all.
This can get quite complicated fast.

**Comment by ruslo** [comment]
> And my public headers are depending on Eigen3 headers

Yes, in the workaround I've mentioned public headers should not depend on third party headers (at least it's a simplest solution).

> But somehow one must ensure that the Eigen3 headers are installed into a private prefix

Why private? It should be public too. For example your library do include Eigen3 headers by `#include <Eigen/something>` and you put everything to `private/Eigen`. In this case when somebody want to use your library the requirements for header path should be something like `-I/path/to/MyLib/public -I/path/to/MyLib/private`. Now imaging that user want to use Eigen3 too and adds it by `-I/path/to/my/Eigen`. As a result build will have `-I/path/to/my/Eigen -I/path/to/MyLib/public -I/path/to/MyLib/private` hence MyLib public will **not use** Eigen3 headers from `-I/path/to/MyLib/private`. Compile errors and ODR vioaltions guaranteed in a long term.

Note that with GIT_SELF feature everything installed into one directory (as usual in Hunter), there is no `private/Eigen` directory.

**Comment by ithron** [comment]
Ah, yes, you're right, Eigen should be public then.

But I don't understand how the GIT_SELF feature solves the problem.
I would include Eigen using `hunter_add_package(Eigen)` and `find_package(Eigen3)`. 
But when I do `make install` all files form MyLib getting installed, but Eigen's not.

**Comment by ruslo** [comment]
> I would include Eigen using hunter_add_package(Eigen) and find_package(Eigen3).
But when I do make install all files form MyLib getting installed, but Eigen's not.

With GIT_SELF you're packing and installing MyLib to the same place where Eigen3 installed, i.e. there should be `hunter_add_package(MyLib)` somewhere. In this case both Eigen3 and MyLib installed to the same folder (internal Hunter root directory). Since GIT_SELF introduce new package MyLib then Config-ID will be new too, hence there is no other third parties in Hunter internal install directory.

Probably it's easier to try it first than explain:
* https://docs.hunter.sh/en/latest/user-guides/hunter-user/git-self.html

**Comment by epvbergen** [comment]
GIT_SELF seems an excellent solution when the project is a library. However, for typical packages that deliver a few executables, a few dynamically loaded libraries with late-loaded plugins and some runtime files, it's important that packages can both install themselves to HUNTER_ROOT as well as to a final CMAKE_INSTALL_PREFIX. 

At some point you need to create a .zip archive, bundle, Debian package or Windows installer.

Perhaps hunterized packages should define both an install 'target' (to $HUNTER_ROOT/_Base/x/y/z/Install) as well a a 'deploy' target to an arbitrary prefix, so that whole dependency hierarchies incl. run time files can be deployed without having to implement deployment for everything from top level. It's the packages that know best what run time support they need installed after all.

I really don't want to cheat the way eg. qt-camera does by finding gstreamer plugins based on a QT_ROOT variable, and hacking together globs and copies.

**Comment by ruslo** [comment]
> At some point you need to create a .zip archive, bundle, Debian package or Windows installer.

For Windows installer or macOS bundle, you should use [fixup_bundle](https://cmake.org/cmake/help/latest/module/BundleUtilities.html). As for other package managers I don't think Hunter should be used there, see:
- https://docs.hunter.sh/en/latest/overview/compatibility.html
- https://docs.hunter.sh/en/latest/faq/interaction-with-other-package-managers.html
